test/e2e/auth/node_authn.go

```diff
@@ -0,0 +1,101 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/master/ports"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:NodeAuthenticator]", func() {
+
+	f := framework.NewDefaultFramework("node-authn")
+	var ns string
+	var nodeName string
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		nodeList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(nodeList.Items)).NotTo(Equal(0))
+		nodeName = nodeList.Items[0].Name
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Get("default", metav1.GetOptions{})
+		Expect(len(sa.Secrets)).NotTo(Equal(0))
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	framework.ConformanceIt("The kubelet's main port 10250 should fail with no credentials", func() {
+		config, err := framework.LoadConfig()
+		Expect(err).NotTo(HaveOccurred())
+
+		c, err := clientset.NewForConfig(restclient.AnonymousClientConfig(config))
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+		defer cancel()
+
+		_, err = c.CoreV1().RESTClient().Get().
+			Context(ctx).
+			Resource("nodes").
+			SubResource("proxy").
+			Name(fmt.Sprintf("%v:%v", nodeName, ports.KubeletPort)).
+			Suffix("stats/summary").
+			Do().Raw()
+		Expect(apierrors.IsUnauthorized(err)).Should(Equal(true))
+	})
+
+	It("Can delegate ServiceAccount tokens to the API server", func() {
+		By("Create a new ServiceAccount for testing")
+		falseValue := false
+		newSA := &v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Name: "newSA"}, AutomountServiceAccountToken: &falseValue}
+		sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Create(newSA)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Get the new created ServiceAccount")
+		saSecret, err := f.ClientSet.CoreV1().Secrets(ns).Get(sa.Secrets[0].Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Delegate new ServiceAccount tokens to the API server")
+		config, err := framework.LoadConfig()
+		Expect(err).NotTo(HaveOccurred())
+		token, err := base64.StdEncoding.DecodeString(string(saSecret.Data["token"]))
+		Expect(err).NotTo(HaveOccurred())
+		caData, err := base64.StdEncoding.DecodeString(string(saSecret.Data["ca.crt"]))
+		Expect(err).NotTo(HaveOccurred())
+		config = restclient.AnonymousClientConfig(config)
+		config.BearerToken = string(token)
+		config.TLSClientConfig = restclient.TLSClientConfig{
+			CAData: caData,
+		}
+		c, err := clientset.NewForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+		_, err = c.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```

Comment:
Will this actually work when you enforce RBAC? Don't you have to create RBAC rules to make the SA able to access the kubelet proxy? Also, I'd expect you to talk to the kubelet directly:

```go
_, err = c.CoreV1().RESTClient().Get().
			Context(ctx).
			Resource("nodes").
			SubResource("proxy").
			Name(fmt.Sprintf("%v:%v", nodeName, ports.KubeletPort)).
			Suffix("metrics").
			Do().Raw()
```